File: BFS-DFS/9663.py
# 시간복잡도 문제로 c++로 풂
n = int(input())

def possible(board, y):
    for i in range(y):
        if board[i] == board[y] or abs(board[y] - board[i]) == y - i:
            return False
    return True

def set_queen(board, row):
    if row == n:
        global result
        result += 1
        return
    for i in range(n):
        board[row] = i
        if possible(board, row) == True:
            set_queen(board, row + 1)

result = 0        
board = [0] * n 
set_queen(board, 0)
print(result)

File: BFS-DFS/9663.cpp
#include <iostream>
#include <cstring>

int result = 0;
int n;
int board[15];

bool possible(int row)
{
    for (int i = 0; i < row; i++)
    {
        if (board[i] == board[row] || (abs(board[i] - board[row]) == row - i))
            return false;
    }
    return true;
}

void    set_queen(int row)
{
    if (row == n)
    {
        result++;
        return ;
    }
    for (int i = 0; i < n; i++)
    {
        board[row] = i;
        if (possible(row))
            set_queen(row + 1);
    }   
}

int main()
{
    std::cin >> n;
    std::memset(board, 0, sizeof(int) * 15);
    set_queen(0);
    std::cout << result;
    return (0);
}